src/components/ExtensionCard.tsx

```diff
@@ -21,64 +21,74 @@ interface IProps {
+    [ExtensionAction.UPDATE]: ExtensionAction.UNINSTALL,
+    [ExtensionAction.UNINSTALL]: ExtensionAction.INSTALL,
+    [ExtensionAction.INSTALL]: ExtensionAction.UNINSTALL,
+} as const;
+
 export default function ExtensionCard(props: IProps) {
     const {
         extension: { name, lang, versionName, installed, hasUpdate, obsolete, pkgName, iconUrl, isNsfw },
         notifyInstall,
     } = props;
-    const [installedState, setInstalledState] = useState<string>(() => {
+    const [installedState, setInstalledState] = useState<InstalledStates>(() => {
         if (obsolete) {
-            return 'obsolete';
+            return InstalledState.OBSOLETE;
         }
         if (hasUpdate) {
-            return 'update';
+            return InstalledState.UPDATE;
         }
-        return installed ? 'uninstall' : 'install';
+        return installed ? InstalledState.UNINSTALL : InstalledState.INSTALL;
     });
 
     const [serverAddress] = useLocalStorage<String>('serverBaseURL', '');
     const [useCache] = useLocalStorage<boolean>('useCache', true);
 
     const langPress = lang === 'all' ? 'All' : lang.toUpperCase();
 
-    function install() {
-        setInstalledState('installing');
-        client.get(`/api/v1/extension/install/${pkgName}`).then(() => {
-            setInstalledState('uninstall');
-            notifyInstall();
-        });
-    }
+    const requestExtensionAction = async (action: ExtensionAction): Promise<void> => {
+        const nextAction = EXTENSION_ACTION_TO_NEXT_ACTION_MAP[action];
+        const state = EXTENSION_ACTION_TO_STATE_MAP[action];
 
-    function update() {
-        setInstalledState('updating');
-        client.get(`/api/v1/extension/update/${pkgName}`).then(() => {
-            setInstalledState('uninstall');
-            notifyInstall();
-        });
-    }
-
-    function uninstall() {
-        setInstalledState('uninstalling');
-        client.get(`/api/v1/extension/uninstall/${pkgName}`).then(() => {
-            // setInstalledState('install');
-            notifyInstall();
-        });
-    }
+        setInstalledState(state);
+        await client.get(`/api/v1/extension/${action.toLowerCase()}/${pkgName}`);
+        setInstalledState(nextAction);
+        notifyInstall();
+    };
 
     function handleButtonClick() {
         switch (installedState) {
-            case 'install':
-                install();
-                break;
-            case 'update':
-                update();
-                break;
-            case 'obsolete':
-                uninstall();
-                setTimeout(() => window.location.reload(), 3000);
+            case ExtensionAction.INSTALL:
+            case ExtensionAction.UPDATE:
+            case ExtensionAction.UNINSTALL:
+                requestExtensionAction(installedState).catch(() => {});
                 break;
-            case 'uninstall':
-                uninstall();
+            case ExtensionState.OBSOLETE:
```

Comment:
@schroda I wanted to know if this is the same mechanism used in the tachiyomi app. Last time this happend it was confusing to me. When I clicked on the obsolete button it started to uninstall the extension. Should we change this behaviour so that the user knows what they are doing by clicking on this?

Reply:
no idea how that works in tachiyomi, never used that app

> Maybe just saying that the extension is obsolete as a label and then giving the uninstall button should work. I remember tachiyomi app doing that once, long time ago

another option could also be to add a group "Obsolote" (or some other name) in the "Source" screen and have the "action" be "Uninstall"

but this can/should be done in a separate pr

Reply:
Agreed, this should be taken up in another PR, rest of the things look good